kinesis/src/main/scala/com.snowplowanalytics.snowplow.collectors.scalastream/sinks/KinesisSink.scala

```diff
@@ -234,13 +231,14 @@ class KinesisSink private (
             }
           case Failure(f) =>
             log.error("Writing failed with error:", f)
-            val errorMessage = {
-              if (retriesLeft > 1)
-                s"Retrying writing batch to SQS queue ${sqs.sqsBufferName} in $nextBackoff milliseconds..."
-              else
-                s"Sending batch to Kinesis stream $streamName in $nextBackoff milliseconds..."
+
+            if (retriesLeft > 1) {
+              log.error(s"Retrying writing batch to SQS queue ${sqs.sqsBufferName} in $nextBackoff milliseconds...")
+            } else {
+              outage = true
```

Comment:
Yes, it is possible that Kinesis is available at this time. I had the same concern as you raise, but ultimately decided that if we have fallen back to SQS it is because Kinesis cannot cope, and in that case SQS being unavailable should be considered an outage.

Also, if Kinesis is still available, successful write attempts to it with other batches would immediately set the flag back to `false`. Does that make sense?

Reply:
> I think that this only complicates the code

I agree with that, sadly.

> and does not really achieve much

Let's talk about what it achieves!

OK, the situation we are now discussing is:
- Kinesis is healthy
- SQS is unhealthy
- We have a surge in traffic, so we have exceeded the kinesis rate limit.

And the problem is:
- The sink health endpoint might return true or false depending on if we get lucky when we query it.

There is another situation which I'm not completely happy with at the moment:
- Kinesis is unhealthy
- SQS is healthy

In that situation the sink health endpoint never reports as unhealthy.  Because all events get successfully sunk to sqs.  But in that situation we probably want to fail over to the backup pipeline in a different region (don't we?).


Is it better to report as unhealthy when there is a problem with kinesis OR sqs?  Maybe this is the desired behaviour:
- If kinesis has had _any errors in the last 60 seconds_ then the health endpoint returns false (not including hitting the rate limit)
- If sqs has had _any errors in the last 60 seconds_ then the health endpoint returns false

You might need to keep track of timestamps to make this possible:

```scala
def isHealthy: Boolean =
  now - lastSeenKinesisError < 60.seconds || now - lastSeenSqsError < 60.seconds
```

WDYT?  Is this achieving anything yet, or still adding unnecessary complications?

Reply:
I really want to land on something simple and unambiguous which will be readily understood by anyone looking at some monitoring built on top of the `sinks-health` endpoint, and also for anyone dealing with the codebase later.

How about: if both Kinesis and SQS are enabled, then we don't care about Kinesis at all, as long as SQS is healthy and all events are being safely stored?

In that case, what happens when only Kinesis is down (or if there is a surge and Kinesis is struggling)? The `sinks-health` endpoint will show `200` but users can (and should) also have monitoring on the SQS traffic volume. As soon as that one spikes, it's a sign that Kinesis is having trouble: but surge protection is kicking in and the collector as a whole is still sinking data, hence can be considered in good health.

Reply:
> I really want to land on something simple and unambiguous

Me too.

> How about: if both Kinesis and SQS are enabled, then we don't care about Kinesis at all, as long as SQS is healthy and all events are being safely stored?

I agree your latest proposal guarantees the events are safely stored.  It also means events are never re-routed to the backup pipeline - I want to explore if that is a problem or not.  Is there any _extra_ reason that we want to re-route events to the backup region, apart from keeping them safe?  E.g. does the backup pipeline have enrich and loaders enabled?

Also..... if there is suddenly a lot of SQS traffic, does that trigger the autoscaling group to increase the number of collectors?  If the sqs traffic is happening because kinesis is down (not because of a surge) then it is not the correct reaction to scale up the collectors.  Do we need a change in our monitoring to behave better in this situation?

Reply:
> It also means events are never re-routed to the backup pipeline 

They will be re-routed if both Kinesis if SQS is down and we try to write to it for any reason (either Kinesis is down also, or it is overwhelmed by surging traffic).

> if there is suddenly a lot of SQS traffic, does that trigger the autoscaling group to increase the number of collectors?

The autoscaling is based on CPU utilisation, so it will autoscale is there is a lot of traffic to the collector, regardless of whether it's going to Kinesis, SQS or both. This is already the case, so we won't be introducing any changes. The protection against a massive EC2 bill here is putting a limit on how many EC2 instances the ASG can deploy.